Reset flag on DataSourceConfig

kaapstorm opened the pull request:
## Technical Summary

Context:
* [Data Source rebuilding getting stuck for indefinite time after adding new Case Properties/Columns](https://dimagi.atlassian.net/browse/QA-7573)
* [Reset Data source build/rebuild flag when migrated instead of rebuild](https://dimagi.atlassian.net/browse/SC-4399)

This change determines which rebuild flags to set on `DataSourceConfiguration.meta.build`, based on whether the data source will automatically be rebuilt.

:fish: :tropical_fish: :blowfish: 

## Safety Assurance

### Safety story

* Tested locally:
  + Creating a new data source
  + Making changes that require rebuilds
  + Making changes that do not require rebuilds
  + Making changes when a rebuild is pending

### Automated test coverage

Yes.

### QA Plan

No.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

kaapstorm committed:
Update comment

kaapstorm committed:
Simplify `if`s across calling and called code.

kaapstorm committed:
Refactor: Model handles state flags

Follows up https://github.com/dimagi/commcare-hq/pull/35633/files#r1941251407

kaapstorm committed:
`pillow_utils` preps config the same as `views`

mkangia commented on corehq/apps/userreports/pillow_utils.py:
```diff
@@ -103,6 +103,10 @@ def rebuild_table(adapter, diffs=None):
         return
 
+    if not config.is_static:
+        config.set_build_queued(reset_init_fin=False)
```
and we will reset awaiting to False if the table is being migrated and not rebuilt by pillows?

kaapstorm replied:
Ah! Thank you, this is the feedback I'm looking for.

If the table is migrated and not rebuilt, are these the correct flags? 
```python
    config.meta.build.awaiting = False

    # Reset these, or leave unchanged?
    config.meta.build.initiated = None
    config.meta.build.finished = False  # Or `True`?
```

mkangia replied:
I can confirm that this is correct and is the one that blocks users from taking more action.
`config.meta.build.awaiting = False`

I am not sure how and where others are relevant.

kaapstorm replied:
0ebcf5dbec9

mkangia replied:
Sorry I missed this earlier,
I believe at this point, `awaiting` is already set to True since that is done in the view so we should not need to queue it again here. The bug was that the flag was not getting removed.

mkangia reviewed (commented):
Nice clean up @kaapstorm :+1:

kaapstorm committed:
Set awaiting = False when build not required

mkangia commented:
thanks for the updates @kaapstorm 
The changes look sufficient to me. Is this ready for review?

kaapstorm commented:
> Is this ready for review?

Nearly. I'm going to rebase, and I'd like to do some testing. It should be ready later today.

mkangia commented:
If you are interested in exploring this a little further, I noticed that we could instead also choose to not set the flag if the data source won't be rebuilt anyway.
This would be done [here](https://github.com/dimagi/commcare-hq/blob/6552af589634e398d15f7c19a3db678ef2ad3742/corehq/apps/userreports/ui/forms.py#L292) by checking if the table will be rebuilt, similarly as done [here](https://github.com/dimagi/commcare-hq/blob/923e329ef9887d0192df0c3bac38a238ddbffbd4/corehq/apps/userreports/pillow_utils.py#L55).
This PR, the way it is currently, is still an improvement and this further improvement can be done as a follow up as well.

kaapstorm commented:
> we could instead also choose to not set the flag if the data source won't be rebuilt anyway.

Good idea!

mkangia commented on corehq/apps/userreports/models.py:
```diff
@@ -753,6 +754,52 @@ def rebuild_failed(self):
+        self.meta.build.awaiting = True
+        if reset_init_fin:
+            self.meta.build.initiated = None
+            self.meta.build.finished = False
```
I noticed this only yesterday when looking at some code,
does this need to also reset initiated_in_place & finished_in_place?
I believe those are just the "in_place" equivalents when "Rebuild in place" is used, and I believe `set_build_queued` will be called for that operation too.

mkangia replied:
Okay, reading more it seems you just retained the implementation as it was.
so, its possible rebuild in place has a bug?

kaapstorm replied:
Yeah, I didn't change the functionality.

I haven't tested this scenario yet, but I suspect that if there is a bug it wouldn't prevent rebuilds, it would just tell the user than an in-place rebuild is finished when it's actually still in progress. The other possibility is that we don't actually need to reset these values.

I don't think this stuff is covered by unit tests, and I'd like to fix that.

mkangia commented on corehq/apps/userreports/models.py:
```diff
@@ -753,6 +754,52 @@ def rebuild_failed(self):
+            self.meta.build.initiated = None
+            self.meta.build.finished = False
+
+    def save_build_started(self, *, in_place=False):
```
nit: I am confused by the asterik here. Is it meant to absorb any arguments?

mkangia replied:
Thanks,
I had forgotten about that :+1:

mkangia commented on corehq/apps/userreports/pillow_utils.py:
```diff
@@ -78,6 +78,8 @@ def migrate_tables_with_logging(engine, diffs, table_names, adapters_by_table):
     for table in table_names:
         adapter = adapters_by_table[table]
+        if not adapter.config.is_static:
+            adapter.config.save_build_not_required()
```
:+1: 

Can this be done after the actual migration, in the for loop below?
And I believe this won't be needed anymore once the follow up PR is done, which will not set the flag only in case the data source does not need a rebuild.

mkangia reviewed (commented):
Sorry I missed something earlier, so just confirming on that, one commit might not be needed.

kaapstorm committed:
isort

kaapstorm committed:
Not Google Analytics

... or Salesforce Analytics, or AWS Analytics, or Oracle Analytics, or Adobe
Analytics, or Tableau CRM Analytics, or IBM Cognos Analytics, or ...

kaapstorm committed:
Revert "`pillow_utils` preps config the same as `views`"

This reverts commit 95a5c400928596641efa2e344a9275bab848dffa.

kaapstorm committed:
Revert "Set awaiting = False when build not required"

This reverts commit 466d6e9d8bb287eb64f860e42876e7d296909152.

kaapstorm committed:
DataSourceConfiguration determines rebuild flags

kaapstorm committed:
Test is upside down

kaapstorm committed:
Tests

kaapstorm committed:
Flake8

kaapstorm committed:
Spell "data source" the same as other places

kaapstorm commented:
I've reverted two previous commits, and fixed the way flags are set in commit https://github.com/dimagi/commcare-hq/pull/36186/commits/5db57238ea311ed7355c294f0b883d87547e35c2.

I added some tests in commit https://github.com/dimagi/commcare-hq/pull/36186/commits/34042dc46a6c2e91fa07a3559e3eceae0a7ca686.

The rest of the commits are just nits I found along the way. Still easier to review by commit.

kaapstorm committed:
Bootstrap 5 Migration - Rebuilt diffs

kaapstorm committed:
Tweak behavior: Set awaiting if build has started

mkangia commented on corehq/apps/userreports/tasks.py:
```diff
@@ -260,7 +255,7 @@ def _iteratively_build_table(config, resume_helper=None, in_place=False, limit=-
 @task(serializer='pickle', queue=UCR_CELERY_QUEUE, ignore_result=True)
 def delete_data_source_task(domain, config_id):
-    from corehq.apps.userreports.views import delete_data_source_shared
+    from .views import delete_data_source_shared
```
I didn't know that isort was configured or itself prefers the dot notation for modules within the same app.

mkangia commented on corehq/apps/userreports/views.py:
```diff
@@ -1204,12 +1204,16 @@ def post(self, request, *args, **kwargs):
+                        'This data source will be built / rebuilt automatically by CommCare HQ.'
+                    ))
+                else:
+                    message_list.append(_('This change does not require a rebuild.'))
```
nit: Can we either skip it or say that "HQ will not perform a data source rebuild for the changes made. Please initiate one manually if needed."

It appears that a rebuild might be needed at times, for instance, adding a nullable column is not considered to be rebuildable, so the user would need to rebuild separately if they want to populate existing rows
https://github.com/dimagi/commcare-hq/blob/0b590d2e093fe694e54d8a8ebdca0c79246c97c5/corehq/apps/userreports/alembic_diffs.py#L113

kaapstorm replied:
Good call. 7253e9aef29

mkangia commented on corehq/apps/userreports/models.py:
```diff
@@ -755,12 +758,35 @@ def rebuild_failed(self):
+    def set_rebuild_flags(self):
+        from .rebuild import get_table_diffs
+
+        if not self.rebuild_awaiting_or_in_progress:
```
The logic after this looks solid.
I am a bit unsure about this check here.

If there is a build awaiting, we don't need to set flags again, makes sense.
if there is a build in progress though, should not we set `awaiting` again?
Additionally, `rebuild_failed` is quite unreliable. I will be adding ticket to remove its usage soon.

mkangia replied:
I see this was updated later to rely only on awaiting, so this is good.

mkangia commented on corehq/apps/userreports/models.py:
```diff
@@ -755,12 +758,35 @@ def rebuild_failed(self):
+                diff.table_name
+                for diff in diffs
+                if diff.type in DiffTypes.TYPES_FOR_REBUILD
+            }
```
nit: I assume this is duplicated, so can it be extracted.
Fine to ignore if not duplicated.

kaapstorm replied:
Yeah, there is a function that returns the same dictionary. 0f8db3b979e

mkangia replied:
okay, i thought the whole loop was same as in pillows.

mkangia reviewed (commented):
Thanks @kaapstorm 
This looks great. Just couple of minor things.

kaapstorm committed:
Use function instead of dict comprehension

kaapstorm committed:
Suggest manual rebuild of necessary

mkangia commented:
oh wow, auto merge is dangerous.

Didn't realize the description was pretty outdated and it does not talk about how this has been tested.
Can you update it and share that this was tested?
